pkg/syncer/storage/storage_pv_controller.go

```diff
@@ -0,0 +1,210 @@
+	getDownstreamPersistentVolumeFromNamespaceLocator func(namespaceLocator shared.NamespaceLocator) (runtime.Object, error)
+	getDownstreamPersistentVolumeClaim                func(persistentVolumeClaimName, persistentVolumeClaimNamespace string) (runtime.Object, error)
+	getUpstreamPersistentVolume                       func(clusterName logicalcluster.Name, persistentVolumeName string) (runtime.Object, error)
+}
+
+// NewPersistentVolumeSyncer returns a new storage persistent volume syncer controller.
+func NewPersistentVolumeSyncer(
+	syncerLogger logr.Logger,
+	syncTargetWorkspace logicalcluster.Name,
+	syncTargetName, syncTargetKey string,
+	upstreamClient kcpdynamic.ClusterInterface,
+	downstreamClient dynamic.Interface,
+	downstreamKubeClient *kubernetes.Clientset,
+	upstreamInformers kcpdynamicinformer.DynamicSharedInformerFactory,
+	downstreamInformers dynamicinformer.DynamicSharedInformerFactory,
+	syncTargetUID types.UID,
+) (*PersistentVolumeController, error) {
+
+	c := &PersistentVolumeController{
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), persistentVolumeControllerName),
+		downstreamClient:  downstreamClient,
+		upstreamInformers: upstreamInformers,
+		upstreamClient:    upstreamClient,
+		syncTarget: syncTargetSpec{
+			name:      syncTargetName,
+			workspace: syncTargetWorkspace,
+			uid:       syncTargetUID,
+			key:       syncTargetKey,
+		},
+		updateDownstreamPersistentVolumeClaim: func(ctx context.Context, persistentVolumeClaim *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+			persistentVolumeClaim, err := downstreamKubeClient.CoreV1().PersistentVolumeClaims(persistentVolumeClaim.Namespace).Update(ctx, persistentVolumeClaim, metav1.UpdateOptions{})
+			if err != nil {
+				return nil, err
+			}
+
+			return persistentVolumeClaim, nil
+		},
+		getDownstreamPersistentVolumeClaim: func(persistentVolumeClaimName, persistentVolumeClaimNamespace string) (runtime.Object, error) {
+			return downstreamInformers.ForResource(persistentVolumeClaimSchemeGroupVersion).Lister().ByNamespace(persistentVolumeClaimNamespace).Get(persistentVolumeClaimName)
+		},
+		getUpstreamPersistentVolume: func(clusterName logicalcluster.Name, persistentVolumeName string) (runtime.Object, error) {
+			// We don't need the namespace since the PV is cluster scoped.
+			return upstreamInformers.ForResource(persistentVolumeSchemeGroupVersion).Lister().ByCluster(clusterName).Get(persistentVolumeName)
+		},
+		getDownstreamPersistentVolumeFromNamespaceLocator: func(namespaceLocator shared.NamespaceLocator) (runtime.Object, error) {
+			namespaceLocatorJSONBytes, err := json.Marshal(namespaceLocator)
+			if err != nil {
+				return nil, err
+			}
+			persistentVolumes, err := downstreamInformers.ForResource(persistentVolumeSchemeGroupVersion).Informer().GetIndexer().ByIndex(byNamespaceLocatorIndexName, string(namespaceLocatorJSONBytes))
```

Comment:
We should use the indices helpers here.

Reply:
Under `pkg/indexers` - here: https://github.com/kcp-dev/kcp/blob/5916c3cd79ca30b3ed6e8a8983d10034f1a926e2/pkg/indexers/indexers.go#L74-L87